app/javascript/controllers/pathogen/tabs_controller.js

```diff
@@ -0,0 +1,647 @@
+   * Debounced to prevent excessive updates during rapid tab switching (e.g., arrow keys)
+   * @private
+   * @param {number} index - The tab index
+   * @param {Function} method - The history method to use (default: history.pushState)
+   * @returns {void}
+   */
+  #updateUrlHash(index, method = history.pushState) {
+    // Clear any pending update
+    if (this.#hashUpdateTimeout) {
+      clearTimeout(this.#hashUpdateTimeout);
+    }
+
+    // Debounce the update by 100ms
+    // This prevents excessive URL updates during rapid arrow key navigation
+    this.#hashUpdateTimeout = setTimeout(() => {
+      try {
+        const hash = this.#getTabHash(index);
+        const url = new URL(window.location.href);
+        // Clear stale tab query params so subsequent submissions don't carry outdated values
+        url.searchParams.delete("tab");
+        url.hash = hash;
+
+        // Use Turbo's history API to properly integrate with Turbo navigation
+        // This ensures back/forward navigation works correctly with Turbo Drive
+        Turbo.session.history.update(method, url, uuidv4());
+      } catch (error) {
+        console.error("[pathogen--tabs] Error updating URL hash:", error);
+      } finally {
+        this.#hashUpdateTimeout = null;
+      }
+    }, 100);
+  }
+
+  /**
+   * Gets the tab index from the current URL hash
+   * @private
+   * @returns {number} The tab index, or -1 if not found
+   */
+  #getTabIndexFromHash() {
+    try {
+      const hash = window.location.hash.slice(1); // Remove the '#'
+      if (!hash) {
+        return -1;
+      }
+
+      // Ensure targets are available (defensive check for morph scenarios)
+      if (!this.hasTabTarget || !this.hasPanelTarget) {
+        return -1;
+      }
+
+      // Try to find tab by ID
+      const tabIndex = this.tabTargets.findIndex(
+        (tab) => tab && tab.id === hash,
+      );
+      if (tabIndex !== -1) {
+        return tabIndex;
+      }
+
+      // Try to find panel by ID
+      const panelIndex = this.panelTargets.findIndex(
+        (panel) => panel && panel.id === hash,
+      );
+      if (panelIndex !== -1) {
+        return panelIndex;
+      }
+
+      // Try to parse as tab-{index} format
+      const match = hash.match(/^tab-(\d+)$/);
+      if (match) {
+        const index = parseInt(match[1], 10);
+        if (index >= 0 && index < this.tabTargets.length) {
+          return index;
+        }
+      }
+
+      return -1;
+    } catch (error) {
+      console.error(
+        "[pathogen--tabs] Error getting tab index from hash:",
+        error,
+      );
+      return -1;
+    }
+  }
+
+  /**
+   * Handles browser hash change events (back/forward navigation)
+   * @private
+   * @returns {void}
+   */
+  #handleHashChange() {
+    try {
+      const hashIndex = this.#getTabIndexFromHash();
+      if (hashIndex !== -1) {
+        // Use replaceState when responding to hash change to avoid creating additional history entries
```

Comment:
Suggestion (blocking): Can you update this comment, it isn't quite accurate. I think it should be something like.

```
// Use replaceState when hash is not present, to avoid creating additional history entries. (e.g. on initialization hash is not present and is set on initial connect, we don't want two history entries in this case).
```